File: python/src/redtonomous/tools/web.py
import requests


def fetch_url(url: str, method: str = "GET", body: str | None = None, headers: dict | None = None) -> str:
    try:
        h = {"User-Agent": "redtonomous/0.1", **(headers or {})}
        resp = requests.request(method, url, data=body, headers=h, timeout=30)
        content_type = resp.headers.get("content-type", "")
        text = resp.text
        if len(text) > 8000:
            text = text[:8000] + f"\n… (truncated, total {len(resp.content)} bytes)"
        return f"STATUS: {resp.status_code}\nCONTENT-TYPE: {content_type}\n\n{text}"
    except Exception as e:
        return f"ERROR: {e}"

File: typescript/src/tools/web.ts
export async function fetchUrl(args: {
  url: string;
  method?: string;
  body?: string;
  headers?: Record<string, string>;
}): Promise<string> {
  const { default: fetch } = await import("node-fetch");
  try {
    const resp = await fetch(args.url, {
      method: args.method ?? "GET",
      headers: { "User-Agent": "redtonomous/0.1", ...(args.headers ?? {}) },
      body: args.body ?? undefined,
    });
    let text = await resp.text();
    if (text.length > 8000) text = text.slice(0, 8000) + `\n… (truncated, ${text.length} bytes total)`;
    return `STATUS: ${resp.status}\nCONTENT-TYPE: ${resp.headers.get("content-type") ?? ""}\n\n${text}`;
  } catch (e) {
    return `ERROR: ${String(e)}`;
  }
}
